ENH: calculation of mean in spatial.transform.Rotation

pmla opened the pull request:
<!-- 
Thanks for contributing a pull request! Please ensure that
your PR satisfies the checklist before submitting:
http://scipy.github.io/devdocs/dev/contributor/development_workflow.html#checklist-before-submitting-a-pr

Also, please name and describe your PR as you would write a
commit message:
https://docs.scipy.org/doc/numpy/dev/gitwash/development_workflow.html#writing-the-commit-message

Note that we are a team of volunteers; we appreciate your
patience during the review process.

Again, thanks for contributing!
-->

#### Reference issue
<!--Example: Closes gh-WXYZ.-->
Implements the feature suggestion in #10862

#### What does this implement/fix?
<!--Please explain your changes.-->
This implements calculation of the mean and variance of rotations. The metric used is the chordal L2 distance.

#### Additional information
<!--Any additional information you think is important.-->
See #10862 for further discussion.

nmayorov commented on scipy/spatial/transform/rotation.py:
```diff
@@ -1465,6 +1466,42 @@ def magnitude(self):
+        mean = self.__class__(v[:, index], normalized=True)
+
+        if return_variance:
+            variance = np.sum((self.as_dcm() - mean.as_dcm())**2 / len(self))
```
I'm not convinced by that computation.

1. Why is the variance of dcm elements is chosen as the measure of variance?
2. At least it should be explained in the docstring.

What if we choose the RMS of rotation magnitudes from the mean rotation to given rotations instead? At least it has the same units as rotation.

nmayorov replied:
I see that here you basically computes the objective function.

It needs some more considerations.

nmayorov commented on scipy/spatial/transform/rotation.py:
```diff
@@ -1465,6 +1466,42 @@ def magnitude(self):
+        >>> r.mean().as_quat()
+        array([0.96505487, 0.14713643, 0.17081499, 0.13357476])
+        """
+
```
Remove blank line?

pmla replied:
Yep.  Done.

pmla commented on scipy/spatial/transform/rotation.py:
```diff
@@ -1465,6 +1466,42 @@ def magnitude(self):
+        Returns
+        -------
+        mean : `Rotation` instance
+            Object containing the mean of the rotations in the current
```
Fixed.

pmla commented on scipy/spatial/transform/rotation.py:
```diff
@@ -1465,6 +1466,42 @@ def magnitude(self):
+        Examples
+        --------
+        >>> from scipy.spatial.transform import Rotation as R
+        >>> r = R.random(10, np.random.RandomState(0))
```
Thanks. Fixed.

pmla commented on scipy/spatial/transform/rotation.py:
```diff
@@ -1465,6 +1466,42 @@ def magnitude(self):
+
+        quat = self._quat.reshape((len(self), 4))
+        K = np.dot(quat.T, quat)
+        l, v = np.linalg.eig(K)
```
Nice.  I didn't know about this function.

pmla commented on scipy/spatial/transform/rotation.py:
```diff
@@ -1465,6 +1466,42 @@ def magnitude(self):
+        quat = self._quat.reshape((len(self), 4))
+        K = np.dot(quat.T, quat)
+        l, v = np.linalg.eig(K)
+        index = np.argmax(l)
```
Done.

pmla commented on scipy/spatial/transform/rotation.py:
```diff
@@ -1465,6 +1466,42 @@ def magnitude(self):
+        array([0.96505487, 0.14713643, 0.17081499, 0.13357476])
+        """
+
+        quat = self._quat.reshape((len(self), 4))
```
Fixed.

nmayorov commented on scipy/spatial/transform/rotation.py:
```diff
@@ -1465,6 +1466,42 @@ def magnitude(self):
+
+        Notes
+        -----
+        The mean used is the chordal L2 mean (also called the projected or
```
What I wanted to say --- perhaps it is worth adding this formulation, because it says more about the actual algorithm.

nmayorov replied:
Will you add a more specific mathematical description? I think it is worth knowing that it essentially works in DCM space.

pmla replied:
> Will you add a more specific mathematical description? I think it is worth knowing that it essentially works in DCM space.

I missed this comment.  Does the current description suffice?

pmla commented:
I have added an option to compute a weighted average as you suggest.  I have implemented the weighted variance using this reference: https://en.wikipedia.org/wiki/Weighted_arithmetic_mean#Weighted_sample_variance

I have added a test to check that doubling a weight is equivalent to including a rotation twice.

nmayorov commented on scipy/spatial/transform/rotation.py:
```diff
@@ -1465,6 +1466,63 @@ def magnitude(self):
+                                 "equal to number of rotations, got "
+                                 "{} values and {} rotations.".format(
+                                    weights.shape[0], len(self)))
+            if np.min(weights) < 0:
```
np.any(weights < 0) is slightly more to the point )

pmla replied:
Nice.

nmayorov commented on scipy/spatial/transform/rotation.py:
```diff
@@ -1465,6 +1466,63 @@ def magnitude(self):
+
+        # K is quadratic in quat, so we take the square root of the weights
+        quat = self._quat * np.sqrt(weights)[:, None]
+        K = np.dot(quat.T, quat)
```
Maybe `K = np.dot(weights * self.quat_.T, self.quat_)` without compute `sqrt` from weights.

pmla replied:
Very nice.

pmla commented:
Let `d(a, b)` be a distance function on rotations.  Then
```m = argmin_{x in SO(3)} \sum\limits_i d^2(x, r_i)```
is the L2-mean, and
```v = \sum\limits_i d^2(m, r_i)```
is the variance.

The distance function we are using here is
```d(a, b) = ((a.as_dcm() - b.as_dcm())**2).sum()```
We use this to calculate the mean, and the variance is also calculated using this function.

> What if we choose the RMS of rotation magnitudes from the mean rotation to given rotations instead? At least it has the same units as rotation.

If we used `d(a, b) = (a * b.inv()).magnitude()` as the distance function, then the variance would be the RMS of rotation magnitudes.  This distance function is more complicated to implement though, which is why I want to start with the chordal distance.

pmla commented:
I really like this paper:
https://users.cecs.anu.edu.au/~hartley/Papers/PDF/Hartley-Trumpf:Rotation-averaging:IJCV.pdf
Sections 4 + 5 provide an excellent description of the different distance functions and their associated means.

pmla commented on scipy/spatial/transform/rotation.py:
```diff
@@ -1465,6 +1466,42 @@ def magnitude(self):
         else:
             return angles
 
+    def mean(self, return_variance=False):
```
We can call it the average if you prefer.  I think that we should use the formal name in the notes section though.

nmayorov replied:
I leave it to you to decide, both names are completely fine.

pmla commented on scipy/spatial/transform/rotation.py:
```diff
@@ -1465,6 +1466,42 @@ def magnitude(self):
             return angles
 
+    def mean(self, return_variance=False):
+        """Get the mean of the rotation(s).
```
Yes.  I was just using the same notation in the docstrings of other methods (inv, magnitude etc.)

nmayorov replied:
So keep it like this? I think a question might arise --- "What is the mean of a single rotation?"

pmla replied:
Oh - I didn't understand what you meant before.  I have changed it to `rotations`.

nmayorov commented:
> I really like this paper:
https://users.cecs.anu.edu.au/~hartley/Papers/PDF/Hartley-Trumpf:Rotation-averaging:IJCV.pdf
Sections 4 + 5 provide an excellent description of the different distance functions and their associated means.

I have read (not very thoroughly), it looks like the best source on this subject. Do you plan on working on other means?

From what I've figured L2 quaternion mean is easy, if we can choose right signs for quaternions. For other means we need to solve the optimization problem iteratively.

> Let d(a, b) be a distance function on rotations. Then
m = argmin_{x in SO(3)} \sum\limits_i d^2(x, r_i)
is the L2-mean, and
v = \sum\limits_i d^2(m, r_i)
is the variance.

> The distance function we are using here is
d(a, b) = ((a.as_dcm() - b.as_dcm())**2).sum()
We use this to calculate the mean, and the variance is also calculated using this function.

Yes, it makes sense mathematically. But how does a user interpret this variance? Maybe I'm wrong, but it doesn't look useful at all from a practical standpoint.

I believe that returning weighted RMS of angular spread from the estimated rotation will be more useful. What do you think?

pmla commented:
> From what I've figured L2 quaternion mean is easy, if we can choose right signs for quaternions.

Figuring out the signs is easy if the rotations are not too far apart (see Theorem 7, pg 19).  In the general case finding the correct signs is a hard problem though - I have a solution which I am writing up in a paper but not published yet.  It is a combinatorial problem so the calculation is slower.

> Do you plan on working on other means?

Yes.  I had planned to implement this one first, and then add keywords to select other means.  At first we can just throw an exception if the rotations are too far apart.

> I believe that returning weighted RMS of angular spread from the estimated rotation will be more useful. What do you think?

If you think of a rotation in terms of its DCM representation the variance is easy to interpret.  I don't like the idea of calculating the mean and variance using different distance functions.

nmayorov commented:
> If you think of a rotation in terms of its DCM representation the variance is easy to interpret. I don't like the idea of calculating the mean and variance using different distance functions.

Mathematically I'm all in favor for that. But for example if we get `variance = 0.01` --- what does it tell us, is it big or small? According to the paper the chordal distance is equal to `2 * sqrt(2) * sin(angle /2)`, so it `~ sqrt(2) * angle` for small angles (well, we even have a `sqrt(2)` factor).

On the other hand, if we get `0.1` radians RMS --- now we have some understanding of how rotations are distributed around the mean.

All in all I'm fine with returning `variance` as it is now, but I suspect it's not going to be very useful. At least it should be explained in the docstring --- the distance choice and that variance is computed using this distance. Do you see other ways to make the situation more clear, perhaps return 2 variances (chordal and natural geodetic)?

Other things to consider:

1. Should we return variance or standard deviation? Of course both are equivalent, but s.d. is always more convenient to reason about.
2. Can we compute variance/s.d. of the estimate (i.e. accuracy)? I guess it's going to be difficult and it is not well defined when the rotations are far apart.

pmla commented:
How about we use the standard deviation rather than the variance, and apply the transformation `theta = 2 * np.arcsin(stddev / (2 * np.sqrt(2)))`.  Then we have a measure of the spread as an angle in the range 0-120 degrees.

nmayorov commented:
> How about we use the standard deviation rather than the variance, and apply the transformation theta = 2 * np.arcsin(stddev / (2 * np.sqrt(2))). Then we have a measure of the spread as an angle in the range 0-120 degrees.

> I think this meets interpretability requirements. We might have to come up with a different name though to make it clear that this isn't the unmodified standard deviation.

It's too tricky, I think computing angular spread RMS directly is just better.

---------

I suggest to keep the variance computations as it is now, but explain it in the docstring. Perhaps turn it into standard deviation.

However, I can hardly see why anyone would be interested in this variance. What do you think?

nmayorov commented on scipy/spatial/transform/rotation.py:
```diff
@@ -1465,6 +1466,61 @@ def magnitude(self):
+        Examples
+        --------
+        >>> from scipy.spatial.transform import Rotation as R
+        >>> r = R.random(10, random_state=0)
```
I think this example is not very meaningful, moreover the resulting mean quaternion can be returned multiplied by -1 and the doctest will fail (this happened).

I suggest generating a set of rotation through Euler angles filling some narrow ranges uniformly. And then represent the mean as Euler angles too. This way we can see that the average has angles somewhere in the middle.

pmla replied:
I have added a very simple example along the lines you suggest.

pmla commented:
> However, I can hardly see why anyone would be interested in this variance. What do you think?

I have removed the variance feature altogether.  It is too complicated to do in a way which is both intuitive and makes mathematical sense.  We can add it in a later PR if the need arises.

nmayorov commented on scipy/spatial/transform/rotation.py:
```diff
@@ -1465,6 +1466,55 @@ def magnitude(self):
+        mean : `Rotation` instance
+            Object containing the mean of the rotations in the current
+            instance.
+        weights : array_like shape (N,), optional
```
weights belong to Parameters.

pmla replied:
Thanks.  That was silly.

nmayorov commented on scipy/spatial/transform/rotation.py:
```diff
@@ -1465,6 +1466,60 @@ def magnitude(self):
+                                    weights.shape[0], len(self)))
+            if np.any(weights < 0):
+                raise ValueError("`weights` must be non-negative.")
+        weights = weights / np.sum(weights)
```
We don't need to normalize weights to compute the mean I suppose.

pmla replied:
Removed.

nmayorov commented on scipy/spatial/transform/rotation.py:
```diff
@@ -1465,6 +1466,60 @@ def magnitude(self):
+        --------
+        >>> from scipy.spatial.transform import Rotation as R
+        >>> r = R.from_euler('zyx', [
+        ... [0, 0, 0],
```
Alignment is not very good, try formatting first in an editor and then add >>> and ...

pmla replied:
I think it is better now.

nmayorov commented on scipy/spatial/transform/rotation.py:
```diff
@@ -1465,6 +1466,60 @@ def magnitude(self):
+        ... [1, 0, 0],
+        ... [0, 1, 0],
+        ... [0, 0, 1]], degrees=True)
+        >>> np.degrees(r.mean().as_euler('zyx'))
```
as_euler('zyx', degrees=True)

pmla replied:
Fixed.

nmayorov commented on scipy/spatial/transform/rotation.py:
```diff
@@ -1465,6 +1466,60 @@ def magnitude(self):
+        The mean used is the chordal L2 mean (also called the projected or
+        induced arithmetic mean). If `p` is a set of rotations with mean `m`,
+        then `m` is the rotation which minimizes
+        `((p.as_dcm() - m.as_dcm())**2).sum()`.
```
Use double ticks for everything here.

pmla replied:
Fixed.

nmayorov commented on scipy/spatial/transform/rotation.py:
```diff
@@ -1465,6 +1466,58 @@ def magnitude(self):
+        The mean used is the chordal L2 mean (also called the projected or
+        induced arithmetic mean). If ``p`` is a set of rotations with mean
+        ``m``, then ``m`` is the rotation which minimizes
+        ``((p.as_dcm() - m.as_dcm())**2).sum()``.
```
Add ``weights`` as multiplies to the formula.

pmla replied:
Done.

nmayorov commented:
Looks excellent, have nothing to add except one note about docstring.

nmayorov commented on scipy/spatial/transform/rotation.py:
```diff
@@ -1465,6 +1466,58 @@ def magnitude(self):
+        The mean used is the chordal L2 mean (also called the projected or
+        induced arithmetic mean). If ``p`` is a set of rotations with mean
+        ``m``, then ``m`` is the rotation which minimizes
+        ``(weights[:, None, None] * (p.as_dcm() - m.as_dcm())**2).sum()``.
```
I updated this line --- weights should multiply the squared difference (checked with another paper).

pmla replied:
What a silly mistake.

nmayorov commented:
Merging after the tests will pass.

nmayorov commented:
@pmla please add notes about rotation groups, reduction and mean into https://github.com/scipy/scipy/wiki/Release-note-entries-for-SciPy-1.4.0#scipyspatial-improvements

And do come back with new ideas and pull requests.

pmla commented:
Thanks again for your help and guidance on this @nmayorov.

I have added this string to the release notes:
```scipy.spatial.transform.Rotation now includes functions for calculation of a mean rotation, generation of the 3D rotation groups, and reduction of rotations with rotational symmetries.```
